### PR TITLE
jinja2.contextfilter has been renamed to pass_context

### PR DESCRIPTION
**Pull request by olemoign**

### CODE REVIEW

**Comment by stevepiercy** [comment]
Test build fails: https://github.com/Pylons/pyramid_jinja2/pull/159/checks?check_run_id=2566704927

Where did you see this change in Jinja2? I couldn't find it in the change log.

https://jinja2docs.readthedocs.io/en/stable/changelog.html

**Comment by olemoign** [comment]
> I reckon we need to drop Python 2.7 and 3.5 at minimum to get tests to pass.

Is this something that you would be willing to do? If so, I can take care of it.

> There are other things that need fixing, too.

Yes, looks like there are a few incompatibilities with pyramid 2, I can check as well.

**Comment by stevepiercy** [comment]
@olemoign yes, please proceed. Thank you!

**Comment by olemoign** [comment]
I can see that compatibility is tested with pyramid 1.3 (last release 2012), I feel we could drop 1.3 and 1.4 and start with 1.5 (last release 2016)?

**Comment by mmerickel** [comment]
I'm not clear what changes are being made that would require dropping the testing for old pyramid versions. I'm ok with it if a new feature requires it but otherwise I don't see the point.

Definitely dropping old python versions should be done.

**Comment by stevepiercy** [comment]
[Pyramid 1.3a1 was the first version to add Python 3 support](https://docs.pylonsproject.org/projects/pyramid/en/latest/changes.html#a1-2011-12-09), so it and later versions should be kept in the test bed.

File: tox.ini
Comment (stevepiercy) [inline tox.ini:5]:
```suggestion
    py39-cover,coverage,
```

File: tox.ini
Comment (stevepiercy) [inline tox.ini:4]:
```suggestion
    py39-pyramid{13,14,15,16,17,18,110,20},
```

File: tox.ini
Comment (stevepiercy) [reply tox.ini:5]: We previously ran coverage on Python 2 (`py2-cover`), then Python 3 (`py3-cover`), then combined them in a third tox environment (`coverage`). This should be simplified to run coverage once on a Python 3 environment and report on it, instead of two environments.

**Review by stevepiercy — CHANGES_REQUESTED** [review]

**Comment by stevepiercy** [comment]
Pyramid 2.0 removed the deprecated scaffolds, and pyramid_layout uses a scaffold when running tests, so they currently fail.

@mmerickel should we switch to using the [cookiecutter](https://github.com/pylons/pyramid-cookiecutter-starter) going forward?

Would this be a compelling reason to drop older versions of Pyramid?

**Comment by olemoign** [comment]
Yeah, I saw that. All tests are passing except for this. I don't super know how cookiecutter works, so I was planning to do the switch on Sunday.
I'm re-adding the lint test that was commented. Isort is good, Flake8 in progress, I'll manage Black in the end.

**Comment by olemoign** [comment]
Btw, the setup.py references the Repoze license at an url that doesn't exist anymore (http://www.repoze.org/LICENSE.txt). Is there a copy of it somewhere else? Is https://github.com/repoze/repoze.who/blob/master/LICENSE.txt or https://www.openhub.net/licenses/repoze-public-license good? Pyramid also references the same url in its setup.py.

**Comment by stevepiercy** [comment]
Perhaps there is something in the [cookiecutter tests](https://github.com/Pylons/pyramid-cookiecutter-starter/blob/latest/tests/test_it.py#L116-L160) that can be reused? Alternatively we could use its rendered files. The [Pyramid docs](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/project.html) have details about how to create a Pyramid project.

I am OK with switching to pytest, too.

For the license URL, please use https://pylonsproject.org/license.html

**Comment by stevepiercy** [comment]
Looking good so far. Coverage and Lint tox envs fail at the moment.

**Comment by olemoign** [comment]
Almost done. I've converted the scaffold to a cookiecutter. I've just spent an hour trying to understand why the translations weren't working, turns out ... the .mo files did not contain the same strings as the .po. :smile: 
This fixes the coverage.

Linting will always fail in cookiecutter on the lines with templating. I think we could ignore it, considering how little code there is inside. The easiest way to ignore it, would be to move the cookiecutter one level up. Touching on this, I think it would make sense to move demo and tests one level up as well. And I'm not sure what's the purpose of having a demo and a cookiecutter which are very very similar (demo is an even simpler version of the cookiecutter).

**Comment by stevepiercy** [comment]
> Linting will always fail in cookiecutter on the lines with templating. I think we could ignore it, considering how little code there is inside. The easiest way to ignore it, would be to move the cookiecutter one level up. Touching on this, I think it would make sense to move demo and tests one level up as well.

Indeed, we should follow the structure of Pyramid.

```
.
├── demo (or cookiecutter, see below)
├── docs
├── src
│   └── pyramid
└── tests
```

> And I'm not sure what's the purpose of having a demo and a cookiecutter which are very very similar (demo is an even simpler version of the cookiecutter).

I'm confused. It looks like you made the directory `cookiecutter`, which is based on `pyramid-cookiecutter-starter` and incorporates the i18n feature from the original scaffold-based demo. The stuff in the directory `cookiecutter` is a demo app, correct? And do you mean you don't see the point of maintaining two demos, one in `demo` and another in `cookiecutter`?

File: MANIFEST.in
Comment (stevepiercy) [inline MANIFEST.in:10]: Why should `.github` be included? I see that Pyramid grafts it, too, but it's only used for CI and not really useful for distribution on a developer's machine. It's fine to leave it, I'm just curious.

File: pyramid_jinja2/__init__.py
Comment (stevepiercy) [inline pyramid_jinja2/__init__.py:10]: Something ain't right. isort or black should not be merging standard library imports with package imports. Was this a manual sort?

File: pyramid_jinja2/__init__.py
Comment (stevepiercy) [inline pyramid_jinja2/__init__.py:126]: In the future, please isolate automated formatting of code to separate PR and as a final step. No one wants to review a bajillion changes to find the ones that matter.

If it is at all possible, please revert any formatting changes for now.

**Review by stevepiercy — CHANGES_REQUESTED** [review]
I aborted my review.

In the future, please isolate automated formatting of code to separate PR and as a final step. No one wants to review a bajillion automated changes to find the ones that matter.

If it is at all possible, please revert any formatting changes for now.

File: .github/workflows/ci-windows.yml
Comment (olemoign) [reply .github/workflows/ci-windows.yml:24]: No other reason that there was separate files before. I can merge them.

File: .gitignore
Comment (olemoign) [reply .gitignore:31]: A lot of the changes that I did were to streamline this package with pyramid. This is pyramid's .gitignore.

File: MANIFEST.in
Comment (olemoign) [reply MANIFEST.in:10]: Same comment than before, I reused pyramid's file. I agree with your point.

File: pyramid_jinja2/__init__.py
Comment (olemoign) [reply pyramid_jinja2/__init__.py:10]: I 100% agree with your comment. This is not the way I would sort this. BUT, once again, I used pyramid's isort config and this is the way imports are sorted there. Here for example: https://github.com/Pylons/pyramid/blob/master/src/pyramid/view.py

File: pyramid_jinja2/__init__.py
Comment (olemoign) [reply pyramid_jinja2/__init__.py:126]: Yes I expected the PR to be too big. But to pass the lint test, you have to lint the code.
I'm going to separate cookiecutter and the black formatting in different PRs.

**Comment by olemoign** [comment]
> I'm confused. It looks like you made the directory `cookiecutter`, which is based on `pyramid-cookiecutter-starter` and incorporates the i18n feature from the original scaffold-based demo. The stuff in the directory `cookiecutter` is a demo app, correct? And do you mean you don't see the point of maintaining two demos, one in `demo` and another in `cookiecutter`?

Before, there were `demo` and `scaffold` folders, now  `demo` and `cookiecutter`. I modified the existing scaffold to fit with cookiecutter and yes, the end result is basically the `pyramid-cookiecutter-starter` jinja2 version with added i18n. So yes, I would call that a demo app. :)
What I'm wondering is the use of having both (demo and scaffold/cookiecutter), when in my mind their use is basically the same.

File: .github/workflows/ci-windows.yml
Comment (stevepiercy) [reply .github/workflows/ci-windows.yml:24]: Ah, so it is. Maybe @bertjwregeer has a reason or preference? Let's leave them separate unless he says otherwise.

**Comment by olemoign** [comment]
I've separated cookiecutter and linting, I'll submit PRs after this one is finished.

File: .gitignore
Comment (stevepiercy) [reply .gitignore:31]: Sure, but we shouldn't just copy-pasta. We should only add necessary bits.

File: MANIFEST.in
Comment (stevepiercy) [reply MANIFEST.in]: I based my opinion on https://packaging.python.org/guides/using-manifest-in/. I don't think it is needed.

File: pyramid_jinja2/__init__.py
Comment (stevepiercy) [reply pyramid_jinja2/__init__.py]: Thank you! I appreciate it.

File: pyramid_jinja2/__init__.py
Comment (stevepiercy) [reply pyramid_jinja2/__init__.py]: Ah, yes, that's right. There may have been a reason for Pyramid's isort config. I remember it was driving me nuts in Deform when I used it as a model. I changed it to make more sense to me and to get black and isort to play nice with each other.

https://github.com/Pylons/deform/blob/main/pyproject.toml

File: MANIFEST.in
Comment (olemoign) [reply MANIFEST.in]: This will cause problems with the `check-manifest` lint command but I'm not sure it's super important.

**Comment by olemoign** [comment]
(The only test that still fails is coverage, because the scaffold is not tested. Considering the other PRs, I consider this a pass.)

**Comment by stevepiercy** [comment]
> What I'm wondering is the use of having both (demo and scaffold/cookiecutter), when in my mind their use is basically the same.

I did a little archeology to try to figure this out.

https://docs.pylonsproject.org/projects/pyramid_jinja2/en/latest/changes.html

## 1.0 (2011-05-12)
- Added demo app just to visualize i18n work

## 1.3 (2011-12-14)
- Make scaffolding compatible with Pyramid 1.3a2+.

The latter converted it from a paster template to a Pyramid scaffold. I think that `demo` was completely forgotten. I am OK with removing it, especially since you've brought in i18n.

The old pcreate scaffold was pretty cool in that it would set up translations and i18n, and I think it's cool that you have maintained that feature in the cookiecutter. We'll have to update the docs, once the dust settles.

https://docs.pylonsproject.org/projects/pyramid_jinja2/en/latest/index.html#pcreate-template-i18n

**Comment by stevepiercy** [comment]
OK, thank you! I'll pick this up again later today and review it more carefully.

File: MANIFEST.in
Comment (stevepiercy) [reply MANIFEST.in]: Correct, I'm picking nits. We could try `prune .github`, the inverse of `graft`.

File: .github/workflows/ci-windows.yml
Comment (bertjwregeer) [reply .github/workflows/ci-windows.yml:24]: No, this was just one of the first packages I touched when I added in the Github actions. We should switch this to be in line with the others!

File: MANIFEST.in
Comment (bertjwregeer) [reply MANIFEST.in]: It is grafted because we have historically in our source dists shipped everything that was checked into git. We continue to do the same with our source distributions now. Please do NOT remove the graft of the .github folder. It is part of our source code and it should be in the source distributable.

File: pyramid_jinja2/__init__.py
Comment (bertjwregeer) [reply pyramid_jinja2/__init__.py]: In Pyramid we made the decision to only treat imports within the project different, everything else is marked 3rd party. This is on purpose, @mmerickel may have more info on why.

File: .gitignore
Comment (bertjwregeer) [inline .gitignore:10]: Not a huge fan of cargo culting this and many of the other entries into the `.gitignore` I assume this too came from pyramid?

File: setup.py
Comment (bertjwregeer) [inline setup.py:50]: Please continue to use double quotes. Please make sure this projects implements a pyproject.toml which sets up black to format all the source files. 

Unless you set the quoting to Black's default, quoting should NOT be modified.

**Review by bertjwregeer — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (olemoign) [reply .gitignore:10]: Yes and based on @stevepiercy comments, the final .gitignore is around half the size of pyramid's one.

File: setup.py
Comment (olemoign) [reply setup.py:50]: Disclaimer: I use single quotes in all my projects so I will be biased.

pyramid_jinja2 is currently 90% single quotes and 10% double quotes. pyramid is mainly single quotes from what I can see. pyramid black's configuration has "skip-string-normalization" and I imported it as is (like I said in the disclaimer, I wasn't too sad about it). 
To be honest, I would feel weird having a different quote system than pyramid. (And more work.)
But this can be the moment we turn it around and I convert everything to double quotes.

Comment (stevepiercy) [reply setup.py:50]: Pylons Project projects use (or should use) [Black](https://black.readthedocs.io/en/stable/index.html) to automatically format code. [Black's default for strings is double-quotes](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings).

I think Pyramid 1.10 was the first Pylons Project project to adopt Black on [2018-10-15](https://github.com/Pylons/pyramid/pull/3388). [Deform added Black on 2018-11-12](https://github.com/Pylons/deform/pull/381). This project just hasn't received as much love over the years. We add Black, isort, and other code formatting automation at the speed of volunteers 🐌 .

Comment (stevepiercy) [reply setup.py]: Also to make it easier to merge this PR, we should not run Black on the entire code base in it. We should do that in a separate PR. The few changes made here are fine.

**Comment by olemoign** [comment]
@stevepiercy @bertjwregeer, it seems the last item was to merge the github test files, I've just done it, is there anything else blocking this PR?

**Comment by stevepiercy** [comment]
@olemoign I just approved the run of GHA and CI passes. Thank you!

We will need to review this PR. I'm backlogged until the weekend, but maybe @bertjwregeer or @mmerickel have a few free cycles.

**Comment by olemoign** [comment]
@stevepiercy @bertjwregeer Gentle reminder about this MR. :smile:

**Comment by olemoign** [comment]
@stevepiercy @bertjwregeer This MR has now been finished for more than 3 months ... :( Is there anything I can do to help you?

**Comment by mmerickel** [comment]
sorry for the delay @olemoign :-( great work on this PR, looking to release it now

**Comment by olemoign** [comment]
I got a bit frustrated, but I 100% understand. Life takes priority over unpaid work.
I have two more PRs (lint and scaffold update) that were awaiting this one, I can probably submit them on Monday. Your call if it's worth waiting.

**Comment by mmerickel** [comment]
ok I just started refactoring the source layout and adding lint so I may be taking over that part but I’m interested in what you are thinking for a scaffold update?

**Comment by olemoign** [comment]
:smile: 
I have already done the lint, just moved it out of this PR because it got too big. I can't submit it until tomorrow at the earliest. Maybe it would save you some time? I'm fine with trashing it if you do it.
Regarding the scaffold, I transformed it into a cookiecutter template.

**Comment by mmerickel** [comment]
I did the blacken/lint steps today but I will wait for the cookiecutter prior to release. Thank you @olemoign!

**Pull request merged**
